src/layer_4/tcp.rs

```diff
@@ -49,16 +49,24 @@ pub fn repl<'a, 'b>(
             };
             /* Compute syncookie */
             if let Ok(cookie) = synackcookie::generate(&client_info, &masscanned.synack_key) {
-                if cookie != ackno {
-                    masscanned.log.tcp_drop(tcp_req, client_info);
-                    return None;
-                }
                 client_info.cookie = Some(cookie);
+                if !proto::is_tcb_set(cookie) {
+                    /* First Ack: check syncookie, create tcb */
+                    if cookie != ackno {
+                        masscanned.log.tcp_drop(tcp_req, client_info);
+                        return None;
+                    }
+                    proto::add_tcb(cookie);
+                }
             }
             warn!("ACK to PSH-ACK on port {}", tcp_req.get_destination());
             let payload = tcp_req.payload();
             /* Any answer to upper-layer protocol? */
-            if let Some(repl) = proto::repl(&payload, masscanned, &mut client_info) {
+            let mut payload_repl = None;
+            proto::get_tcb(client_info.cookie.unwrap(), |tcb| {
```

Comment:
If I understand this correctly, there might be some pretty hard lifetimes issues if we were to have `tcb` inside the `client_info`. `client_info` is currently mutable has little restrictions on it, as it is discarded once the processing is over and juste needs to be passed to the next layer.

On the contrary, `tcb` needs to be persistent as it is stored in the hashmap, so the code needs to be in a closure that locks the mutex (what is done in `tcb.rs`). The other option was to lock the MUTEX each time you needed it (what whas done before but was kinda ugly). Puting `tcb` in `client_info` would be a bit weird, because it means that there would be an optional field that only exists within the closure, then needs to be deleted somehow. And you need to have `client_info` exist before and after this.

Reply:
Ok ; let us merge it like that and we'll see later if it needs to be refactorized.